rfcs/0077_name_mangling_ir_modules.md

```diff
@@ -0,0 +1,102 @@
+- `GlobalVar FreshGlobal(const String name, bool add_prefix)` guarantees to return a newly constructed `GlobalVar` that is guaranteed not to conflict by name with other `GlobalVars` generated by the same `NameSupply` object. 
+The functionality of `add_prefix` is as described above.
+
+The name mangling in case of conflict is currently performed by appending `_1_2_3..._n` to the `name` parameter. 
+A potential improvement would be to treat the integer suffix 
+separately from the string prefix. 
+The following text snippet provides the general idea of mangling improvement:
+```
+x = get_next_name("fun12", existing_names=["func12", "func13", "func4"])
+assert x == "func14"
+``` 
+This can be useful to avoid the current confusion caused by having multiple names in form:
+```
+func_1
+func_1_2
+func_1_2_3
+```
+
+Additionally, the `NameSupply` provides static methods to perform mangling on a string. For example, `String Get_name_mangled(const String& module_name, const String& name)` can be called to prefix a module name to a name. The AOT executor frequently calls this utility with predefined names.
+
+```
+NameSupply name_supply = NameSupply::NameSupplyWithPrefix(mod_name);
+..
+..
+// Will perform a lookup based on candidate_name and return a GlobalVar from the cache if found. Otherwise, a new GlobalVar.
+GlobalVar prim_fn_var = name_supply->UniqueGlobalFor(candidate_name_);
+// will always print "mod_name_candidate_name_"
+std::cout << prim_fn_var->name_hint;
+..
+..
+// Will always return a new GlobalVar with "name_hint" mangled.
+GlobalVar prim_fn_gvar = name_supply->FreshGlobal(candidate_name);
+// Will print "mod_name_candidate_name_0_1_2_3..n" depending on many times this name was used with "name_supply".
+std::cout << prim_fn_gvar->name_hint;
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `NameSupply` will be implemented as a class in TVM following the usual design, extending `Object` and `ObjectRef`, respectively.
+As mentioned, the `NameSupply` will contain an internal map of `String -> GlobalVar` used to perform lookups by `UniqueGlobalFor`. Additionally, a map of `String -> Int` will be used to provide unique names for `FreshGlobal`.
+
+Additional refactorings to the compiler will be performed as part of this RFC. For example, the `GlobalVar` deduplication in `TECompilerImpl::LowerInternal` will be removed by changing the signature of `tvm::LowerSchedule`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- There might be cases when creating new `GlobalVars` without using a `NameSupply` might be needed. For example, in `IRModule::FromExprInContext`, the expression might already be annotated with a global symbol.
+We must be sure that the `NameSupply` interacts well with the `global_symbol` attribute. 
```

Comment:
I think since global_symbol is already endemic and a string we just rely on UniqueGlobalFor to recover the GlobalVar.